cirq/circuits/circuit.py

```diff
@@ -884,6 +886,13 @@ def insert(
                                   preserve_moments=True),
             preserve_moments=True))
 
+        common_keys = _get_measurement_keys(
```

Comment:
This is a very expensive check to perform every time there is an insertion... It iterates over every operation in the entire circuit, causing quadratic overhead.

If we're going to do this, we need to cache the set of used measurements and keep that cache up to date.

This is why I originally specified the idea in terms of the Simulator verifying that the circuit has no dupes; keeping the cache up to date will be a pain.

Reply:
Thanks for the review @Strilanc.

I weighed the current option against the simulator. I saw all the circuit/device constraints are checked when the circuit is being built and hence thought this check should also be there to maintain consistency. 

As you can see, I am not checking for measurement_key duplication within a moment in `_get_measurement_keys` method as that is already validated when constructing the moment. This is a slight cut in computation cost, but still I agree it is expensive. 

The other simple way to cut the computation cost is to cache the measurement keys at the moment level. I didn't do it because `Moment` class currently has no member variables like this.

Reply:
Please let me know what do you think!

Reply:
Gentle ping for review.

Reply:
Iterating over the entire circuit every time we add an operation is too inefficient. This has to be fixed before we can merge.

Reply:
Do you prefer the idea of verifying in simulator compared to the proposal of adding a cache at the moment level ?